File: backend/src/models/product.model.ts
import { IProduct } from '@/types';
import { ProductCategory, ProductStatus } from '@/types/product.type';
import mongoose from 'mongoose';

const ProductRecipeItemSchema = new mongoose.Schema(
  {
    ingredientId: { type: mongoose.Schema.Types.ObjectId, ref: 'Ingredient', required: true },
    quantity: { type: Number, required: true },
    unit: { type: String, required: true, trim: true },
  },
  { _id: false }
);

const ProductStoreAvailabilitySchema = new mongoose.Schema(
  {
    storeId: { type: mongoose.Schema.Types.ObjectId, ref: 'Store', required: true },
    status: {
      type: String,
      required: true,
      enum: [ProductStatus.ACTIVE, ProductStatus.INACTIVE, ProductStatus.OUT_OF_STOCK],
    },
  },
  { _id: false }
);

const ProductSchema = new mongoose.Schema<IProduct>(
  {
    nameEmbedding: { type: String, default: null },
    imgEmbedding: { type: String, required: true },
    name: { type: String, required: true, trim: true },
    description: { type: String },
    image: { type: String },
    price: {
      type: Number,
      required: true,
      min: [0.01, 'Price must be greater than 0'],
    },
    category: { type: String, required: true, enum: ProductCategory },
    restaurant: { type: String, trim: true },
    time: { type: String, trim: true },
    rating: { type: Number, default: 0, min: 0, max: 5 },
    reviewCount: { type: Number, default: 0, min: 0 },
    recipe: { type: [ProductRecipeItemSchema], default: [] },
    allergenTags: { type: [String], default: [] },
    healthWarning: { type: String },
    healthTags: { type: [String], default: [] },
    isAvailable: { type: Boolean, default: true },
    isCampaignRunning: { type: Boolean, default: false },
    variationIds: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Variation' }],
    storeAvailability: { type: [ProductStoreAvailabilitySchema], default: [] },
    crossContaminationRisk: { type: Boolean, default: false },
    mayContain: { type: [String], default: [] },
    embedding: { type: [Number], default: [] },
    tags: { type: [String], default: [] },
    operationalNote: { type: String, trim: true },
  },
  {
    timestamps: true,
  }
);

// Indexes
ProductSchema.index({ category: 1 });
ProductSchema.index({ 'storeAvailability.storeId': 1 });
ProductSchema.index({ price: 1 });
ProductSchema.index({ name: 'text', description: 'text' }); // Text search indexing

const ProductModel = mongoose.model<IProduct>('Product', ProductSchema, 'products');

export default ProductModel;

File: mobile/lib/core/models/product_model.dart
class ProductModel {
  final String id;
  final String? storeId;
  final String name;
  final String description;
  final String imageUrl;
  final int price;
  final String category;
  final String? restaurant;
  final String? time;
  final double rating;
  final int reviewCount;
  final List<String> tags;
  final String? healthWarning;
  final List<String> healthTags;
  final int? campaignPrice;
  final bool isAvailable;
  final String status;
  final List<VariantGroupModel>? variants;

  ProductModel({
    required this.id,
    this.storeId,
    required this.name,
    required this.description,
    required this.imageUrl,
    required this.price,
    required this.category,
    this.restaurant,
    this.time,
    this.rating = 0.0,
    this.reviewCount = 0,
    this.tags = const [],
    this.healthWarning,
    this.healthTags = const [],
    this.campaignPrice,
    this.isAvailable = true,
    this.status = 'active',
    this.variants,
  });

  factory ProductModel.fromJson(Map<String, dynamic> json) {
    String imageStr = '';
    final imageJson = json['image'];
    if (imageJson is String) {
      imageStr = imageJson;
    } else if (imageJson is Map<String, dynamic>) {
      imageStr = imageJson['secureUrl'] as String? ?? '';
    }

    return ProductModel(
      id: json['_id'] as String? ?? json['id'] as String? ?? '',
      storeId: json['storeId'] as String?,
      name: json['name'] as String? ?? '',
      description: json['description'] as String? ?? '',
      imageUrl: imageStr,
      price: json['price'] as int? ?? 0,
      category: json['category'] as String? ?? '',
      restaurant: json['restaurant'] as String?,
      time: json['time'] as String?,
      rating: (json['rating'] as num?)?.toDouble() ?? 0.0,
      reviewCount: json['reviewCount'] as int? ?? 0,
      tags: (json['tags'] as List?)?.map((e) => e as String).toList() ?? [],
      healthWarning: json['healthWarning'] as String?,
      healthTags:
          (json['healthTags'] as List?)?.map((e) => e as String).toList() ?? [],
      campaignPrice: json['campaignPrice'] as int?,
      isAvailable: json['isAvailable'] as bool? ?? true,
      status: json['status'] as String? ?? 'active',
      variants: json['variants'] != null
          ? (json['variants'] as List)
                .map(
                  (e) => VariantGroupModel.fromJson(e as Map<String, dynamic>),
                )
                .toList()
          : (json['variationIds'] != null
                ? (json['variationIds'] as List)
                      .map(
                        (e) => VariantGroupModel.fromJson(
                          e as Map<String, dynamic>,
                        ),
                      )
                      .toList()
                : null),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'storeId': storeId,
      'name': name,
      'description': description,
      'image': {'secureUrl': imageUrl},
      'price': price,
      'category': category,
      'restaurant': restaurant,
      'time': time,
      'rating': rating,
      'reviewCount': reviewCount,
      'tags': tags,
      'healthWarning': healthWarning,
      'healthTags': healthTags,
      'campaignPrice': campaignPrice,
      'isAvailable': isAvailable,
      'status': status,
      'variants': variants?.map((e) => e.toJson()).toList(),
    };
  }
}

class VariantGroupModel {
  final String name;
  final bool? required;
  final bool? multiple;
  final int? maxChoices;
  final List<VariantOptionModel> options;

  VariantGroupModel({
    required this.name,
    this.required,
    this.multiple,
    this.maxChoices,
    required this.options,
  });

  factory VariantGroupModel.fromJson(Map<String, dynamic> json) {
    return VariantGroupModel(
      name: json['name'] as String? ?? '',
      required: json['required'] as bool?,
      multiple: json['multiple'] as bool?,
      maxChoices: json['maxChoices'] as int?,
      options: json['options'] != null
          ? (json['options'] as List)
                .map(
                  (e) => VariantOptionModel.fromJson(e as Map<String, dynamic>),
                )
                .toList()
          : const [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'required': required,
      'multiple': multiple,
      'maxChoices': maxChoices,
      'options': options.map((e) => e.toJson()).toList(),
    };
  }
}

class VariantOptionModel {
  final String choice;
  final int extraPrice;

  VariantOptionModel({required this.choice, required this.extraPrice});

  factory VariantOptionModel.fromJson(Map<String, dynamic> json) {
    return VariantOptionModel(
      choice: json['choice'] as String? ?? '',
      extraPrice: json['extraPrice'] as int? ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {'choice': choice, 'extraPrice': extraPrice};
  }
}
